src/terminal/Terminal.cpp

```diff
@@ -1729,4 +1729,40 @@ void Terminal::setHighlightRange(HighlightRange _range)
     highlightRange_ = _range;
     eventListener_.updateHighlights();
 }
+
+void Terminal::pushColorPalette(size_t slot)
+{
+    constexpr size_t MaxColorPaletteSaveStackSize = 10;
+
+    if (slot > MaxColorPaletteSaveStackSize)
+        return;
+
+    if (slot == 0)
+        slot = state_.savedColorPalettes.empty() ? 0 : state_.savedColorPalettes.size() - 1;
+    else
+        slot--;
+
+    if (state_.savedColorPalettes.size() <= slot)
+        state_.savedColorPalettes.resize(slot + 1);
+
+    state_.savedColorPalettes[slot] = state_.colorPalette;
+}
+
+void Terminal::popColorPalette(size_t slot)
+{
+    if (state_.savedColorPalettes.empty())
+        return;
+
+    if (slot == 0)
+        slot = state_.savedColorPalettes.size() - 1;
+    else
+        slot--;
+
+    if (!state_.savedColorPalettes.empty())
+    {
+        state_.colorPalette = move(state_.savedColorPalettes.back());
+        state_.savedColorPalettes.pop_back();
+    }
```

Comment:
Slots 1 through 10 should only restore the color palette and not pop the stack.

> Parameters (integers in the range 1
 >         through 10, since the default 0 will pop) may be used to
  >        restore the palette from the stack without popping.

Reply:
done